working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+them into an existing proposal, or add a new feature specification.
+
+
+## Design principles
+
+We aim to minimize the differences between structs and classes.  As much as
+possible, structs should behave as restrictions of classes, and extension
+structs should behave as further restrictions of structs.  We specifically aim
+to avoid as much as possible having different behaviors for the same concept
+(e.g. differences in scoping).
+
+We also aim to minimize the amount of new syntax required, and to maximize the
+amount of new functionality that we can provide relative to the syntactic real
+estate consumed, and the new cognitive load imposed on users.
+
+## Roadmap
+
+The first section describes the full struct feature.  The second section
+describes the restriction of the full struct to support static wrappers.
+
+## Structs
+
+This section describes how structs look and behave, largely by example.  In many
+cases, alternatives or extensions to the core proposal are described in line.
+Larger extensions are postponed to a later section.
+
+## Structs by example
+
+Structs allow you to define simple classes very compactly.  For example, we
+might model a component vector as follows:
+
+```
+struct Component(int x, int y, int z, int w = 1);
+```
+
+This short definition essentially defines a class `Component`, with four fields,
+and provides a number of convenient methods.
+
+```dart
+void test() {
+  // Call the default constructor, using the default value for w
+  var c1 = Component(0, 0, 0);
+  // Call the default constructor, explicitly passing w
+  var c2 = Component(1, 1, 1, 1);
+
+  assert(c1 == c2); // Equality is defined to be structural
+  assert(c1.hashCode == c2.hashCode); // With correspondind hashCode
+
+  print(c1.debutToString()); // Prints "Component(0, 0, 0, 1)"
+
+  print(c1.toString()); // Prints "struct"
+
+  // Generated copy method
+  var c2 = c1.copyWith(x : 2);
+  assert(c2.x == 2);
+}
+```
+
+Structs may explicitly define constructors, static members, and normal class
+members (except fields), and may implement interfaces.
+
+```dart
+struct Component(int _x, int _y, int _z, int _w = 1)
+  implements Comparable<Component> {
+
+  double get x => _x/_w;
+  double get y => _x/_w;
+  double get z => _z/_w;
+
+  int compareTo(Component other) => throw "TODO: Unimplemented";
+
+  Component.zero() : _x = 0, _y = 0, _z = 0;
+}
+```
+
+With the addition of pattern matching and switches, structs will support closed
+families (algebraic datatypes).
+
+```dart
+abstract struct Operand;
+
+struct ConstantOperand(Value c) extends Operand;
+
+struct IdentifierOperand(Identifier i) extends Operand;
+
+Operand replaceIn(Operand oper, Identifier t, Value c) {
+  return switch (oper) {
+    case ConstantOperand(_) : oper;
+    case IdentifierOperand(var i) where i == t: ConstantOperand(c);
+    case IdentifierOperand(_) : oper;
+  }
+}
+```
+
+## Structs in more detail
```

Comment:
Is there any support for declaring a struct that is immutable?

Reply:
https://github.com/tc39/proposal-record-tuple – it'd be great to layer this on this feature in JS – if it's approved

Reply:
All structs are shallow immutable.  There's no support for deep immutability in this proposal.

Reply:
Filed https://github.com/dart-lang/language/issues/2362 to discuss immutability.